### DISCUSSION TITLE
Filewatchers no longer work

### DISCUSSION THREAD

**Discussion by JokerMartini** [quasarframework/quasar#10012] [CLI - Electron mode]
I'm not sure if this is directly related to the architecture updates made to Quasar or Electron, but I noticed that Filewatchers no longer work in quasar. They all fail to trigger whether you use the node.js filewatcher **fs** or you import and use **import chokidar from 'chokidar';**

They all fail to return events that trigger and print to the console. I was wondering if anyone else is experiencing this problem?
The code below will not trigger or print anything. It will error saying the watcher does not contain a function called 'on'.

I had 2 other developers on reddit try this over a google meet call and they all experienced the same issues both with chokidar and fs packages. Any help would be much appreciated. 

Side note, i can validate these methods work when used outside of the quasar framework.
```
import chokidar from 'chokidar';

  setFileWatcher(filepath) {
    const watcher = chokidar.watch(filepath, {
      depth: 0,
      ignorePermissionErrors: true
    });
    if (watcher) {
      watcher
        .on('ready', () => {
          console.log('FileWatcher monitoring >> ', watcher.getWatched());
        })
        .on('change', path => {
          dispatch('loadConfigFile');
          console.log('FileWatcher changed emitted >>', path);
        })
        .on('error', error => {
          console.error(error);
        });
    }
  },
```

**Comment by justinbeaty** [comment]
@JokerMartini it kind of looks like you're trying to run this code in the render process, which now has `nodeIntegration: false`. This issue is similar to your #9996

As razvan hinted at, you should be able to turn on node integration with:

```
new BrowserWindow({
    // ...
    webPreferences: { // <<<<< this is really important
      nodeIntegration: true,
    }
}
```

**but** I'd be very careful with this. There's a reason why it is (and should be) disabled by default. This is from memory as I haven't used electron mode in a few years, but I'd highly recommend reading more about the webPreferences object.

**Reply by JokerMartini** [reply]
is it a security risk if im making a desktop electron app that is running local on a computer. not through the web browser?

**Reply by justinbeaty** [reply]
@JokerMartini if you’re 100% only loading local content, there’s less risk, but if you’re loading any remote content that could normally trigger an XSS attack, it turns into a RCE attack which could totally takeover the machine. This is good reading: https://www.electronjs.org/docs/tutorial/security#2-do-not-enable-nodejs-integration-for-remote-content

Also to be clear, this isn’t an issue on webpages because webpages don’t use electron.

**Reply by JokerMartini** [reply]
That makes sense. It was a good read. Thanks for taking the time to help read that. 
The app I'm building will run on computers that do not have internet access. It's all internal so no security risks

**Reply by JokerMartini** [reply]
@justinbeaty i did exactly what you said and it still fails to work. I get the following issues. Here is my setup 

**electron-main.js**
```
  mainWindow = new BrowserWindow({
    width: 1000,
    height: 960,
    useContentSize: true,
    webPreferences: {
      contextIsolation: true,
      enableRemoteModule: true,
      nodeIntegration: true,
      preload: path.resolve(__dirname, process.env.QUASAR_ELECTRON_PRELOAD)
    }
  })
```

**electron-preload.js**
```
import { contextBridge } from 'electron'
import chokidar from 'chokidar';

contextBridge.exposeInMainWorld('electron', {
    chokidar
})
```

Then in my actions file where I'm trying to set a filepath to be watched using the filewatcher it fails to return me a proper result. 

**actions.js**
```
export function setFileWatcher({ state, dispatch }) {
  const watcher = window.electron.chokidar.watch(state.configFilepath, {
    depth: 0,
    ignorePermissionErrors: true
  });
  console.log(watcher)
}
```
This screenshot showcases the issue how the object returned in Quasar 2.0 appears to not be actually returning a object that has neither a file in the **maps** or **events** as you can see the **eventsCount** is 3 vs 0 in the later.

![image](https://user-images.githubusercontent.com/1416749/125083355-4870cb80-e096-11eb-9950-b4c45ab796e7.png)

**Reply by JokerMartini** [reply]
@justinbeaty 
**UPDATES**
Ok so I'm slowly getting somewhere but still have a huge issue. If I move the method that creates the actual FSWatcher to a preload js file it works. However I need to have access to this watcher in my state's actions.js file so i can dispatch other methods when the filewatcher notifies me of changes.

**utils.js**
```
export function setFileWatcher(filepath) {
  const watcher = chokidar.watch(filepath, {
    depth: 0,
    ignorePermissionErrors: true
  });
  console.log(watcher)
  return watcher
}
```

**electron-preload.js**
```
import { contextBridge } from 'electron'
import * as utils from '../src/helpers/utils'

contextBridge.exposeInMainWorld('electron', {
    utils
})
```

Now in my **actions.js** file I need to be able to listen to the events of the filewatcher in order to trigger other events in my applications state, however the object returned is not a FSWatcher object anymore... What would you suggest I do here?

**actions.js**
```
export function setFileWatcher({ state, dispatch }) {
  // console.log("set file watcher...");
  const watcher = window.electron.utils.setFileWatcher(state.configFilepath)
  console.log(watcher)
}
```

These are the methods I need to listen to in my state...
 ```
// if (watcher) {
  //   watcher
  //     .on('ready', () => {
  //       console.log('FileWatcher monitoring >> ', watcher.getWatched());
  //     })
  //     .on('change', path => {
  //       dispatch('loadConfigFile');
  //       console.log('FileWatcher changed emitted >>', path);
  //     })
  //     .on('error', error => {
  //       console.error(error);
  //     });
  // }
```

This is what prints when calling console.log(watcher) from the **utils.js** file. (**CORRECT**)
![image](https://user-images.githubusercontent.com/1416749/125085019-237d5800-e098-11eb-8c6b-20df051c4383.png)

This is what prints when calling console.log(watcher) from the **actions.js** file. (**WRONG**)
![image](https://user-images.githubusercontent.com/1416749/125085066-2ed08380-e098-11eb-827e-0cb4ceb28ee4.png)

**Reply by justinbeaty** [reply]
@JokerMartini It's been quite a while since I've used Electron, so I'm very rusty on the subject. But it sounds like the best option is to keep chokidar in preload, and use IPC to communicate between the preload script and render process. The only thing is that the data must be serializable, so you may need to make a simplified object to pass from the preload with just the keys you need.

**Reply by JokerMartini** [reply]
thanks for trying to suggest some solutions justin. I hope that someone or one of the developers of quasar may be able to suggest a solution. All previous versions of the application I made prior to 2.0 allowed for this, it's quite a bummer if this feature is no longer supported in a way that developers can utitlize file watchers.

**Reply by ontwikkelfabriek** [reply]
@JokerMartini 

You could try `watchman` instead of `chokidar` :
https://github.com/facebook/watchman

NodeJs installation:
https://facebook.github.io/watchman/docs/nodejs.html

**Reply by JokerMartini** [reply]
> @JokerMartini
> 
> You could try `watchman` instead of `chokidar` :
> https://github.com/facebook/watchman
> 
> NodeJs installation:
> https://facebook.github.io/watchman/docs/nodejs.html

Why would this suggested package work over the one I mentioned I was using above? I'm guessing the same issues would Persist

**Reply by ontwikkelfabriek** [reply]
@JokerMartini 

> Why would this suggested package work over the one I mentioned I was using above? I'm guessing the same issues would Persist

Really? You can tell already that this will **not** work and will have the **same** issues.  Would probably take you about 15 min to find out...

**Reply by JokerMartini** [reply]
I'm not sure I understand your response. I read over the package you suggested and it seemed to use a similar process that's why I asked if it would work or not, assuming you were more familiar with it 😀

**Reply by ontwikkelfabriek** [reply]
@JokerMartini 
> assuming you were more familiar with it 

Ah no I was just trying to give alternatives. Keep us posted of any progress regarding File Watchers!

**Reply by JokerMartini** [reply]
For sure, it's much appreciated the support and help. I'll give this a test tomorrow and report back.

**Comment by ontwikkelfabriek** [comment]
@JokerMartini 

This one works for me in Qv2:
https://github.com/mikeal/watch

**Comment by iwebxa** [comment]
@JokerMartini  any good news now?

**Reply by JokerMartini** [reply]
sadly i moved away from Electron because the builds are enormous and went to C++ and Qt instead. Electron has so many security issues compared to that of a native C++ application. As for good news, none of the solutions above, using electron for a desktop app that needs to communicate with folders and files on disk has become a huge headache now a days.

**Reply by iwebxa** [reply]
Thank you for your reply.
